### DISCUSSION TITLE
Problem Recording Data from html-button-response

### DISCUSSION THREAD

**Discussion by ncfisherim** [jspsych/jsPsych#1020] [Q&A] [answered]
Hello!

I am new to javascript/programming in general and have been trying to figure out how to record the response to an HTML button response.  I have multiple trials in the timeline variables section and my goal is to connect a different correct answer for certain trials.  So, the part of code I am looking at is under the html-button-response, but it also calls for the correct_response from the timeline variables.  The variable irbtest is the bulk of the experiment so I included the entire variable just in case any of it was relevant.  Here is how I tried to code it (the indentation was changed in certain parts for some reason when posting here):

```
var irbtest = {
    timeline: [
			{
				type: 'animation',
				stimuli: jsPsych.timelineVariable('faces'),
				width: 800,
        frame_time: 750,
        frame_isi: 750,
				choices: jsPsych.NO_KEYS,

			},
			{
        type: "html-button-response",
        stimulus: "Was the second face <strong>darker</strong> or " +
        "<strong>lighter</strong> than the first face?",
        stimulus_height: 1,
        choices: ['DARKER', 'LIGHTER'],
        data: jsPsych.timelineVariable('correct_response'),
        on_finish: function(data){
            if(data.button_pressed===jsPsych.timelineVariable('correct_response')){
                  data.correct = true;
              } else {
                  data.correct = false;
                  console.log(jsPsych.timelineVariable('correct_response'))

            }
        }
      },
      {
        type: 'html-slider-response',
        stimulus: '<p>How confident are you in your previous response?</p>',
        labels: ['Not Confident', 'Neutral', 'Very Confident'],
      }
		],
		timeline_variables: [
			{ faces: ["img/BB1.jpg", "img/WW1.jpg"], correct_response: ["1"] },
      { faces: ["img/BB2.jpg", "img/WW2.jpg"], correct_response: ["0"] },
      { faces: ["img/NN3.jpg", "img/NN5.jpg"], correct_response: ["1"] },
      { faces: ["img/BW3.jpg", "img/WW5.jpg"], correct_response: ["0"] },
      { faces: ["img/WW3.jpg", "img/BW6.jpg"], correct_response: ["1"] },

		],
		randomize_order: true,
    	repetitions: 1
	}
	timeline.push(irbtest);

````

When I do this, at the end it comes up with this:

"0": "1",
"rt": 583.3250000141561,
"stimulus": "Was the second face <strong>darker</strong> or <strong>lighter</strong> than the first face?",
"button_pressed": "1",
"trial_type": "html-button-response",
"trial_index": 15,
"time_elapsed": 32089,
"internal_node_id": "0.0-2.0-1.4",
"subject": "n398zt90l304syn",
"correct": false

So basically it considers it false whether or not I push the correct button.  I am wondering what I am doing wrong in telling the plugin which response is correct?

Thank you for your help in advance!

**Comment by jodeleeuw** [comment, marked as answer]
Hi @ncfisherim ,

This is due to some poor design of the `timelineVariable` function on my part. 

When you use `jsPsych.timelineVariable(...)` inside of a function you need to set the optional second parameter to `true`.

```js
if(data.button_pressed===jsPsych.timelineVariable('correct_response', true)){
                  data.correct = true;
              } else {
                  data.correct = false;
                  console.log(jsPsych.timelineVariable('correct_response', true))

            }
```

We're working on improving this. Check out #788 for some discussion of the issue.

**Reply by ncfisherim** [reply]
Okay, thanks for your help!

For some reason, even with the change it is still unable to detect if the button pressed matches under "if" and always goes to "else", determining the response to be false.  Am I missing something more?  (I copied and pasted exactly what you sent me, and other than that I made no changes to the original yet).

One thing that did change is the console correctly logs which button was pressed; it was not doing that before.

**Reply by becky-gilbert** [reply]
Hi @ncfisherim, I think this is because `correct_response` is stored as an _array_ containing a string (e.g. ["1"]) whereas the value of `button_pressed` in the jsPsych data looks like it's a string (e.g. "1"). So you could try indexing the first (only) element from the `correct_response` array before comparing that to `data.button_pressed`:
```javascript
if (data.button_pressed === jsPsych.timelineVariable('correct_response', true)[0]){
  data.correct = true;
} else {
  data.correct = false;
  console.log(jsPsych.timelineVariable('correct_response', true))
}
```
If that doesn't work then it might be that `data.button_pressed` is an integer rather than a string. You can check by printing the value of `data.button_pressed` to the console at the start of your `on_finish` function.
I hope that helps!

**Reply by ncfisherim** [reply]
Yes!  That helped, thank you both so much!  It turns out the data.button_pressed is a string, so I had to code the correct_response as one, too.  Here is the working code in case anyone else has a similar issue:




![2020-08-31](https://user-images.githubusercontent.com/70032674/91774424-bc109500-ebb6-11ea-8516-8b84f23d20cb.png)
